File: backend/schemas/purchase.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from utils.enums import PurchaseOrderStatus
from utils.quantity_validation import validate_whole_quantity


class PurchaseOrderLineCreate(BaseModel):
    product_id: int
    ordered_quantity: float = Field(gt=0)
    cost_price: float | None = Field(default=None, ge=0)

    @field_validator("ordered_quantity")
    @classmethod
    def _whole_qty(cls, v: float) -> float:
        return validate_whole_quantity(v)


class PurchaseOrderCreate(BaseModel):
    vendor: str = Field(min_length=1, max_length=200)
    responsible_person: str | None = None
    lines: list[PurchaseOrderLineCreate] = Field(min_length=1)


class ReceiptLine(BaseModel):
    line_id: int
    quantity: float = Field(gt=0)

    @field_validator("quantity")
    @classmethod
    def _whole_qty(cls, v: float) -> float:
        return validate_whole_quantity(v)


class ReceiptRequest(BaseModel):
    """Quantities received per line. If omitted, receive all remaining."""
    lines: list[ReceiptLine] | None = None


class PurchaseOrderLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    ordered_quantity: float
    received_quantity: float
    cost_price: float
    total: float
    remaining_to_receive: float


class PurchaseOrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    vendor: str
    responsible_person: str | None
    creation_date: datetime
    status: PurchaseOrderStatus
    source_sales_order_id: int | None
    total_amount: float
    lines: list[PurchaseOrderLineOut]

File: frontend/src/types/purchase.ts
export type PurchaseOrderStatus =
  | "DRAFT"
  | "CONFIRMED"
  | "PARTIALLY_RECEIVED"
  | "RECEIVED"
  | "CANCELLED";

export interface PurchaseOrderLine {
  id: number;
  product_id: number;
  ordered_quantity: number;
  received_quantity: number;
  cost_price: number;
  total: number;
  remaining_to_receive: number;
}

export interface PurchaseOrder {
  id: number;
  vendor: string;
  responsible_person: string | null;
  creation_date: string;
  status: PurchaseOrderStatus;
  source_sales_order_id: number | null;
  total_amount: number;
  lines: PurchaseOrderLine[];
}

export interface PurchaseOrderLineCreate {
  product_id: number;
  ordered_quantity: number;
  cost_price?: number | null;
}

export interface PurchaseOrderCreate {
  vendor: string;
  responsible_person?: string | null;
  lines: PurchaseOrderLineCreate[];
}

export interface ReceiptLineInput {
  line_id: number;
  quantity: number;
}

/** Body for POST /purchase-orders/{id}/receive. Omit `lines` to receive all. */
export interface ReceiptRequest {
  lines?: ReceiptLineInput[];
}
